### PR TITLE
Fix deletion permits flow in RemoteFsTimestampAwareTranslog

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- This PR is making change `RemoteFsTranslog` and `RemoteFsTimestampAwareTranslog` to ensure following:
  - Deletion permits are released when no longer required
  - More than `REMOTE_DELETION_PERMITS` are not available at any given point

### Check List
- [x] Functionality includes testing.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java [hunk 1]
```diff
@@ -240,7 +261,12 @@ public void onResponse(List<BlobMetadata> blobMetadata) {
                         remoteGenerationDeletionPermits.release();
                     }
                 } catch (Exception e) {
+                    logger.error("Exception in trimUnreferencedReaders", e);
                     remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java:265]: When can there be a case now that we would need to release 2 permits during exception thrown now?

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 2]
```diff
@@ -590,7 +590,14 @@ protected void trimUnreferencedReaders(boolean onlyTrimLocal) throws IOException
...
-            deleteRemoteGeneration(generationsToDelete);
+            try {
+                deleteRemoteGeneration(generationsToDelete);
+            } catch (Exception e) {
+                logger.error("Exception in delete generations flow", e);
+                // Release permit that is meant for metadata files and return
+                remoteGenerationDeletionPermits.release();
+                return;
+            }
             translogTransferManager.deleteStaleTranslogMetadataFilesAsync(remoteGenerationDeletionPermits::release);
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:601]: Do we need similar handling for this line as done for deleteRemoteGeneration. We wont need to release permits during exception, but for the logs we can have same behaviour of logging and not throwing it to the caller?

**Review by ashking94 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java (hunk 2 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:601]: `deleteStaleTranslogMetadataFilesAsync ` doesn't throw the exception.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java (hunk 1 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java:265]: This is just a catch-all box for any exception that occurs in the try block.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java [hunk 3]
```diff
@@ -215,21 +215,42 @@ public void onResponse(List<BlobMetadata> blobMetadata) {
...
-                            generationsToBeDeleted,
-                            remoteGenerationDeletionPermits::release
-                        );
+                        try {
+                            translogTransferManager.deleteGenerationAsync(
+                                primaryTermSupplier.getAsLong(),
+                                generationsToBeDeleted,
+                                remoteGenerationDeletionPermits::release
+                            );
+                        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTimestampAwareTranslog.java:224]: I have moved the try-catch block from `translogTransferManager.deleteTranslogFilesAsync` to `translogTransferManager.deleteGenerationAsync`.

**Review by ashking94 — APPROVED** [review]
lgtm

**Review by dbwiddis — APPROVED** [review]

**Pull request merged**